### PR TITLE
Adding configuration file parser for swupd

### LINKED ISSUE #952: Create a swupd config file to define default flags to use on swupd commands
It would be nice to have a config file where a user can set default flags to be used on swupd commands.

For example, forcing --force-insecure-http, --keepcache, --max-retries, --max-parallel-downloads=

We shouldn't allow users to change all flags though. Forcing --nosigcheck, --format, for example, doesn't make sense.

Format of the file to be defined.

### PR DESCRIPTION
**Pull request by castulo**
swupd has many flags that can be used for fine tunning its
functionality. In some cases users may want to include a flag with every
swupd command they run.

This commit gives the ability to provide swupd flags via options in a
configuration file that swupd will read before running the command.

The configuration file is an INI style file tha  can include sections
so users can specify options that should only apply to a specific
command. The biggest advantage of this is to be able to fine tune the
flags that should apply for each command.

For example a user could set a global flag that would apply to every
swupd command, and then turn that flag off for a specific command by
unsetting it in the command's section.

Closes #952 

Signed-off-by: Castulo Martinez <castulo.martinez@intel.com>

### CODE REVIEW

File: src/globals.c [hunk 1]
```diff
@@ -225,13 +248,21 @@ int set_content_url(char *url)
 int set_version_url(char *url)
```
Comment (otaviobp) [inline src/globals.c]: This function is starting to became very confusing. I think it's worth refactoring that and changing how its used so we don't need to have logic in it to decide when to overlap and when not to overlap

Comment (otaviobp) [reply src/globals.c]: The same for set_content_url()

File: configure.ac [hunk 2]
```diff
@@ -239,6 +239,15 @@ AC_ARG_WITH(
 AC_SUBST(USER_ALIAS_PATH, [${user_alias_path}])
 AC_DEFINE_UNQUOTED([USER_ALIAS_PATH], ["${user_alias_path}"], [user-alias-path])
 
+AC_ARG_WITH(
+  [configuration-file],
+  [AS_HELP_STRING([--with-configuration-file=PATH], [Configuration file used to define persistent swupd flags])],
```
[the remark this answers is not in the document]
Comment (castulo) [reply configure.ac]: I think we do. The idea is that the `swupd.config` file gets distributed with Clear (same as swupd.bash) and gets installed in the appropriate directory, in this case /usr/share/clear. The swupd.config file, by default will have all possible sections and flags, but flags will be commented out. This means swupd will always find the config file. So if a user wants to set up a value all they need to do is to go to the file and uncomment whatever flag they want to use, and add the right value to it.

Now, if a user wants to use a different file and from a different location then they can build their own swupd and use the `.configure --with-configuration-file=/my/path/myconfigfile`. So  the purpose of this is not for testing purposes but to provide flexibility to users.

File: configure.ac [hunk 3]
```diff
@@ -239,6 +239,15 @@ AC_ARG_WITH(
...
+  [configuration_file=${withval}],
+  [configuration_file="/usr/share/clear/swupd.config"]
```
[the remark this answers is not in the document]
Comment (castulo) [reply configure.ac]: This file is really not a "template" but the actual config file being read by swupd (although is all commented out at the begging by default) so it doesn't make sense to have it with other defaults.

File: test/functional/usability/usa-config-file.bats [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+}
+
+global_teardown() {
+
+	destroy_test_environment "$TEST_NAME"
+
+}
+
+@test "USA010: Use option without section in the configuration file" {
```
[the remark this answers is not in the document]
Comment (castulo) [reply test/functional/usability/usa-config-file.bats]: it may be slower than unit tests but it is really not slow, all 10 tests run in 2.554s, and functional tests are easier to create. What are the corner cases you think we are missing here?

File: src/lib/strings.h [hunk 5]
```diff
@@ -80,6 +80,12 @@ int strtoi_err_endptr(const char *str, char **endptr, int *value);
  */
 char *str_tolower(const char *str);
 
+/**
+ * Converts a string to a boolean. If the string is "true" then it is converted
```
[the remark this answers is not in the document]
Comment (castulo) [reply src/lib/strings.h]: done

**Comment by miguelinux** [comment]
Hi, I've not review the code, and have a question.

Is this file stateless complain?

I mean it could be on
```
/etc/swupd/config
/runt/swupd/config
/usr/share/swupd/config
```

**Comment by castulo** [comment]
> Hi, I've not review the code, and have a question.
> 
> Is this file stateless complain?
> 
> I mean it could be on
> 
> ```
> /etc/swupd/config
> /runt/swupd/config
> /usr/share/swupd/config
> ```

As of now it is in here -> /usr/share/clear/swupd.config

**Comment by miguelinux** [comment]
> > ```
> > /etc/swupd/config
> > /run/swupd/config
> > /usr/share/swupd/config
> > ```
> 
> As of now it is in here -> /usr/share/clear/swupd.config

Can I ask to be stateless as others packages software in CL?

Just a rule

```
if exist /etc/swupd/config
 open it
else
if exist /run/swupd/config
 open it
else
 open/usr/share/swupd/config
```

File: src/verify.c [hunk 6]
```diff
@@ -1117,3 +1114,21 @@ enum swupd_code verify_main(int argc, char **argv)
 	progress_finish_steps("verify", ret);
 	return ret;
 }
+
+enum swupd_code diagnose_main(int argc, char **argv)
+{
+	int ret = SWUPD_OK;
+
+	string_or_die(&cmdline_option_picky_tree, "/usr");
```
[the remark this answers is not in the document]
Comment (castulo) [reply src/verify.c]: the problem is that when the picky-tree option is being read from the config file we need to allocate memory to store the value or it would be lost. So that means we would need a pointer to a char to read that value instead of const char.

File: configure.ac (hunk 2 above)
[the remark this answers is not in the document]
Comment (castulo) [reply configure.ac]: Actually based on the comments from @miguelinux the file should go to  `/usr/share/defaults/swupd/config` as you mentioned in your next comment @otaviobp  and if the user wants to use it they would have to copy it to either `/etc/swupd` or `/run/swupd` and modify it there. So I guess now it doesn't make sense to have this `--with-configuration-file` option.....

File: configure.ac (hunk 3 above)
[the remark this answers is not in the document]
Comment (castulo) [reply configure.ac]: Ok, change of plans, it will go to  /usr/share/defaults/swupd/config as you requested 👌

**Comment by miguelinux** [comment]
> Ok, change of plans, it will go to  /usr/share/defaults/swupd/config as you requested 👌 
> 

Thanks guys :)

File: test/functional/usability/usa-config-file.bats (hunk 4 above)
[the remark this answers is not in the document]
Comment (otaviobp) [reply test/functional/usability/usa-config-file.bats]: if it's easier to write, go for it.

**Comment by otaviobp** [comment]
btw, I didn't know about /run/. thanks for pointing this.

**Comment by miguelinux** [comment]
> btw, I didn't know about /run/. thanks for pointing this.

I learned it from systemd files ;-)

and is useful when you test config files and don't want to disturb
your admin config from /etc

File: src/clr_bundle_add.c [hunk 7]
```diff
@@ -53,12 +53,15 @@ static void print_help(void)
 
 static const struct option prog_opts[] = {
 	{ "list", no_argument, 0, 'l' },
-	{ "skip-diskspace-check", no_argument, &skip_diskspace_check, 1 },
+	{ "skip-diskspace-check", no_argument, 0, 's' },
```
[the remark this answers is not in the document]
Comment (castulo) [reply src/clr_bundle_add.c]: yeah I don't love it either but I couldn't find another way to handle those long opts from the config file without having an equivalent short opt. The fact that we have a short option for every flag doesn't mean we need to publish/show that short option to the end user. So the end user won't know there is a short option for that flag. So this approach should not become a problem unless we run out of letters for the short options which I don't think is gonna happen, and if it does, we can change this approach then. Or what suggestions do you have for solving this?

[the remark this answers is not in the document]
Comment (castulo) [reply src/clr_bundle_add.c]: One of the things that make this problem hard to solve in a different way is that some options from the command line don't receive an argument, for example `--debug` which only sets the log level at "debug level". However, when using the same value from the config file the same flag will always have an argument, because we could enable a global flag for all commands but we may want to disable it only for one command, so we need to be able to pass a true/false to that flag that would take no argument otherwise.

File: src/globals.c (hunk 1 above)
Comment (castulo) [reply src/globals.c]: done

File: src/clr_bundle_add.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (otaviobp) [reply src/clr_bundle_add.c]: > The fact that we have a short option for every flag doesn't mean we need to publish/show that short option to the end user. 

yes, but the user could use that. And this limits us in the number of parameters that we may have. And it's getting hard to find new short options that makes sense.

> One of the things that make this problem hard to solve in a different way is that some options from the command line don't receive an argument, for example --debug which only sets the log level at "debug level". 

Another option is to create another function that is called by both parse_options and the config parser to set the values instead of making the config parser to call the parse_options.

Not sure if it helps you, but we could use the longindex parameter from getopt_long function to handle long parameters in a switch case too. Another approach I can think of is to set the val in struct options to a value that isn't a valid printed character and avoid adding that char to optstring. According to the manual that's only the value that will be returned by getopt_long() if the function is used. it doesn't affect the short option.

**Comment by castulo** [comment]
@otaviobp this PR is ready to be reviewed again, I have addressed all your concerns.
